packages/core/src/browser/tree/tree-selection.ts

```diff
@@ -5,56 +5,112 @@
  * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
  */
 
-import { injectable, inject } from "inversify";
-import { Event, Emitter, Disposable, SelectionProvider } from "../../common";
-import { ITree, ITreeNode } from "./tree";
-
-export const ITreeSelectionService = Symbol("ITreeSelectionService");
+import { injectable, inject, postConstruct } from 'inversify';
+import { Tree, TreeNode } from './tree';
+import { TreeSelectionState } from './tree-selection-state';
+import { Event, Emitter, Disposable, SelectionProvider } from '../../common';
 
 /**
  * The tree selection service.
  */
-export interface ITreeSelectionService extends Disposable, SelectionProvider<Readonly<ISelectableTreeNode> | undefined> {
+export const TreeSelectionService = Symbol("TreeSelectionService");
+export interface TreeSelectionService extends Disposable, SelectionProvider<ReadonlyArray<Readonly<SelectableTreeNode>>> {
+
+    /**
+     * The tree selection, representing the selected nodes from the tree. If nothing is selected, the
+     * result will be empty.
+     */
+    readonly selectedNodes: ReadonlyArray<Readonly<SelectableTreeNode>>;
+
+    /**
+     * Emitted when the selection has changed in the tree.
+     */
+    readonly onSelectionChanged: Event<ReadonlyArray<Readonly<SelectableTreeNode>>>;
+
+    /**
+     * Registers the given selection into the tree selection service. If the selection state changes after adding the
+     * `selectionOrTreeNode` argument, a selection changed event will be fired. If the argument is a tree node,
+     * a it will be treated as a tree selection with the default selection type.
+     */
+    addSelection(selectionOrTreeNode: TreeSelection | Readonly<SelectableTreeNode>): void;
+
+}
+
+/**
+ * Representation of a tree selection.
+ */
+export interface TreeSelection {
+
     /**
-     * The node selected in the tree. If defined then valid.
-     * Undefined if there is no node selection.
+     * The actual item that has been selected.
      */
-    readonly selectedNode: Readonly<ISelectableTreeNode> | undefined;
+    readonly node: Readonly<SelectableTreeNode>;
+
     /**
-     * Emit when the node selection is changed.
+     * The optional tree selection type. Defaults to `SelectionType.DEFAULT`;
      */
-    readonly onSelectionChanged: Event<Readonly<ISelectableTreeNode> | undefined>;
+    readonly type?: TreeSelection.SelectionType;
+
+}
+
+export namespace TreeSelection {
+
     /**
-     * Select a given node.
-     * If a given node is undefined or invalid then remove the node selection.
+     * Enumeration of selection types.
      */
-    selectNode(node: Readonly<ISelectableTreeNode> | undefined): void;
+    export enum SelectionType {
```

Comment:
Why use options instead of proper methods? I've seen other tree APIs, never something like that, look at VS code tree API or Eclipse.

Reply:
> never something like that

Here you go: https://docs.oracle.com/javase/8/javafx/api/javafx/scene/control/TreeView.html#setSelectionModel-javafx.scene.control.MultipleSelectionModel-

Reply:
Could you clarify what you meant? I see `MultipleSelectionModel` is kind of `SelectionService` and it has proper methods: https://docs.oracle.com/javase/8/javafx/api/javafx/scene/control/MultipleSelectionModel.html
https://docs.oracle.com/javase/8/javafx/api/javafx/scene/control/SelectionModel.html

Reply:
I meant, there other APIs than VSCode's `treeModel`.

Reply:
true, and they don't have one method doing different things depending on options, as you pointed out in https://docs.oracle.com/javase/8/javafx/api/javafx/scene/control/MultipleSelectionModel.html

Reply:
Yes, but they have the [`Item`](https://github.com/Microsoft/vscode/blob/4cc18eca7f620bb464011bb027d699e5b0b5ea84/src/vs/base/parts/tree/browser/treeModel.ts#L241) attached to the selected element doing different things.

Reply:
My point is that instead of one `addSelection`, there should be methods like `select`, `deselect`, `selectRange`, `deselectRange`, `clearSelection` and so on. It is on [TreeModel](https://github.com/Microsoft/vscode/blob/4cc18eca7f620bb464011bb027d699e5b0b5ea84/src/vs/base/parts/tree/browser/treeModel.ts#L873) in VS Code and on [MultipleSelectionModel](https://docs.oracle.com/javase/8/javafx/api/javafx/scene/control/MultipleSelectionModel.html) in Eclipse.

It is nice because one can easily discover them via the content assist and study what they are doing one by one. One does not need to study operation modes of one method. I don't see the reason to convert them to one method having different operation modes. Although, on the implementation side, all of them can call the same method for sure.

Reply:
I see your point. We could do that; would it be sufficient adding these convenient methods to the `TreeModel` but not the `TreeSelectionService`? What's your take on that?

Reply:
it will be fine to have it in `TreeModel`.